### PR TITLE
24 game mechanics

### PR DESCRIPTION
**Pull request by ferinator**

### CODE REVIEW

**Comment by michaelschufi** [comment]
Looks good overall. Can you please run prettier and the linting to fix the errors the build pipeline mentions?

File: game-backend/src/game-engine/health-mechanics/unitUtils.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+import { Unit } from '../../../../darwin-types/game-objects/Unit';
+import { HEALTH_LOSS_RATE } from '../../constants/gameObjects';
+import { getGameObjectsPerType } from '../../helper/gameObjects';
+import produce from '../../helper/produce';
+
+export function updateHealth(state: State) {
+  return produce(state, draft => {
+    const units = getGameObjectsPerType(state, GAME_OBJECT_TYPES.UNIT);
+    units.forEach((unit: Unit) => {
+      unit.health *= HEALTH_LOSS_RATE;
```
Comment (michaelschufi) [inline game-backend/src/game-engine/health-mechanics/unitUtils.ts]: I would actually prefer that the health loss is a flat amount per tick. This line here means that the unit can't reach 0 HP if we don't round.

**Comment by ferinator** [comment]
Test Helper class has been adjusted with new propertie field. Maybe we can take a closer look on how to avoid casting to Unit or Food and instead let TS figure it out.

**Comment by StuckiSimon** [comment]
Could you update your branch and resolve the merge conflict?

File: game-backend/src/game-engine/health-mechanics/unitUtils.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+import { getGameObjectsPerType } from '../../helper/gameObjects';
+import produce from '../../helper/produce';
+
+export default function updateHealth(state: State): State {
+  return produce(state, draft => {
+    const units = getGameObjectsPerType(state, GAME_OBJECT_TYPES.UNIT);
+    units.forEach((gameObject: Unit) => {
+      const unit = draft.objectMap[gameObject.id] as Unit;
+      unit.health -= HEALTH_LOSS_RATE;
+      if (unit.health === 0) {
```
Comment (StuckiSimon) [inline game-backend/src/game-engine/health-mechanics/unitUtils.ts]: Comparing `=== 0` is not really bullet proof. Use lt / gt comparison for such use cases.

File: game-backend/src/game-engine/health-mechanics/unitUtils.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+        delete draft.objectMap[gameObject.id];
+        draft.objectIds = draft.objectIds.filter(id => id !== gameObject.id);
```
Comment (StuckiSimon) [inline game-backend/src/game-engine/health-mechanics/unitUtils.ts]: There is already a helper for this logic, see https://github.com/code-and-conquer/darwin/blob/b9a722f4f34e30d7360e97184f4b00a376f01ea5/game-backend/src/helper/gameObjects.ts#L56

File: game-backend/src/game-engine/health-mechanics/unitUtils.ts [hunk 4]
```diff
@@ -0,0 +1,22 @@
+import { State } from '../../../../darwin-types/State';
+import { GAME_OBJECT_TYPES } from '../../../../darwin-types/game-objects/GameObject';
+import {
+  Unit,
+  HEALTH_LOSS_RATE,
+} from '../../../../darwin-types/game-objects/Unit';
+import { getGameObjectsPerType } from '../../helper/gameObjects';
+import produce from '../../helper/produce';
+
+export default function updateHealth(state: State): State {
```
Comment (StuckiSimon) [inline game-backend/src/game-engine/health-mechanics/unitUtils.ts]: Why is the file called `unitUtils` when the method is called `updateHealth`?

I'd propose to:
- rename the folder `health-mechnics` to `mechanics`
- add a `mechanics/index.ts` which contains the `handleGameMechanics` method which is currently placed in `game-engine/index.ts`

File: game-backend/src/game-engine/health-mechanics/unitUtils.test.ts [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+      .build();
+
+    const expectedState = StateBuilder.buildState()
+      .addUnit({ id: 'test1', x: 1, y: 1, health: 90 })
+      .build();
+
+    const updatedState = updateHealth(state);
+    const unitUpdated: Unit = updatedState.objectMap.test1 as Unit;
+    const unitExpected: Unit = expectedState.objectMap.test1 as Unit;
+    expect(unitUpdated.health).toBe(unitExpected.health);
```
Comment (StuckiSimon) [inline game-backend/src/game-engine/health-mechanics/unitUtils.test.ts]: Why not simply `expect(unitUpdated.health).toBe(90);`?

**Comment by ferinator** [comment]
The removeGameObject seems not to work as expected. The mechancis test seems to work if I use this code inside of updateHealth:
        draft.objectMap = removeGameObject(draft, gameObject.id).objectMap;
        draft.objectIds = removeGameObject(draft, gameObject.id).objectIds;

**Comment by StuckiSimon** [comment]
Check https://github.com/code-and-conquer/darwin/blob/b9a722f4f34e30d7360e97184f4b00a376f01ea5/game-backend/src/game-engine/intent/ConsumeIntent.ts#L38 for an example usage of `removeGameObject`.

**Comment by StuckiSimon** [comment]
This has been covered by https://github.com/code-and-conquer/darwin/pull/60

**Pull request closed without merging**